### DISCUSSION TITLE
RayTracing module: To trace the trajectory in a projectile motion

### DISCUSSION THREAD

**Discussion by sb50nemo** [idaholab/moose#21283] [Q&A Modules: General] [answered]
I want to trace the trajectory of a mass particle in a projectile motion using `RayTracing` module. First of all, is it possible? 
If so, how should I proceed with it? Using `Gravity` kernel. So far from the examples, I could only find rays which are straight lines, is it possible to use RayTracingExodus to output non-linear trajectories?

**Comment by GiudGiud** [comment, marked as answer]
Hello 

It's possible, you can change the direction of each ray/particle at every time step.
It will still be linear on each time step.
Let me see if we can share an example with you

Guillaume

**Reply by sb50nemo** [reply]
Thank you very much, an example would be very helpful and much appreciated.

**Reply by sb50nemo** [reply]
Hii Guillaume, 

Just a kind reminder for you to share an example to use ray tracing module to map the trajectory in a projectile motion.
Many thanks in advance.

--
Somnath

**Reply by GiudGiud** [reply]
Hello

Sorry for the delay, took time to get approval and locate it
https://github.com/loganharbour/moose/tree/rt_mpm is an example of using the RT module to follow particles. Not everything is implemented yet. It seems only generation and contribution to some sort of mass matrix.

I think it s missing the part where the trajectory gets changed. However, having implemented this for monte carlo particle transport (using the RT module) I can help with that.

Guillaume

**Reply by GiudGiud** [reply]
do you have access to that branch? If not we may ping Logan

**Reply by sb50nemo** [reply]
@GiudGiud Many thanks for the example. Yes, I have access to this branch. I will try to implement my problem on this commit and will get back to you.

**Reply by sb50nemo** [reply]
Hello @GiudGiud and @loganharbour, Thank you very much for the example of using ray tracing to follow the mass particles.
I ran the example and I have attached a screenshot of the result, where the bunch of mass particles move to their final location in 7 timesteps, but the ray tracing exodus could only produce the trace of the second last timestep. Also, the trace doesn't change with time.
<img width="1134" alt="Screenshot 2022-10-05 at 6 00 56 PM" src="https://user-images.githubusercontent.com/68417883/194108775-31bb226f-7200-488a-b96f-6220ed8d450d.png">

Here I am attaching the lines I used for `RayTracingExodus`

````
[Outputs]
  exodus = true

  [rays]
    file_base = mpm_rays
    type = RayTracingExodus
    study = study
    execute_on = 'TIMESTEP_END'
  []
[]
````
In short, I have two questions and I would appreciate your help on these:

1. What could I do to get the transient trace of the rays?
2. If my mass particles move in a non-linear fashion, then the transient trace would also be non-linear, right?

Thank you very much for your help, it is highly appreciated, looking forward to your response.

--
Best Regards, 
Somnath

**Reply by GiudGiud** [reply]
2. The trace will be composed of segments. While the trace will not be linear, each segment, corresponding to a timestep, will be linear/straight

1. You can cache more ray tracing information by using these two booleans. I suspect the first one is all you need for this
These options need to be passed to the Ray tracing study, in the user objects 
```
  # Needed to cache trace information for RayTracingMeshOutput
  # always_cache_traces = true
  # Needed to cache Ray data for RayTracingMeshOutput
  # data_on_cache_traces = true
```

**Reply by sb50nemo** [reply]
Thank you for your prompt response, it is very good to know that a non-linear trace is possible. However I am still struggling to get more information stored in the `RayTracingExodus`. I am hereby attaching two blocks from the input file and they yielded the same result as in my previous comment.

```
[UserObjects]
  [study]
    type = MPMBoundingBoxRayStudy
    bbox_min = '0.05 0.8 0'
    bbox_max = '0.3 0.95 0'
    grid_intervals = '5 5'

    displacement = 'disp_x disp_y'
    mass = mass

    ray_kernel_coverage_check = false

    #To output rays
    always_cache_traces = true
    data_on_cache_traces = true

    execute_on = TIMESTEP_END
  []
[]

[Outputs]
  exodus = true

  [rays]
    file_base = mpm_rays
    type = RayTracingExodus
    study = study
    execute_on = 'TIMESTEP_END FINAL'
  []
[]
```

**Reply by GiudGiud** [reply]
do you mind attaching your output file? Maybe I ll understand the problem from there

**Reply by sb50nemo** [reply]
Ofcourse, hereby I am attaching the exodus files and the input file as well.
[mpm_test.zip](https://github.com/idaholab/moose/files/9725688/mpm_test.zip)

**Reply by GiudGiud** [reply]
What kind of work have you done so far on this?

I m worried you're expecting regular kernels to affect the trajectories of the particles, when you really need to be implementing RayKernels for that.

**Reply by sb50nemo** [reply]
On this example, I have added the `[rays]` block in the `Outputs` to use `RayTracingExodus` to trace the motion of the mass particles in the `[study]` block. I didn't try to change the trajectory yet, just tried to visualize it. Should I use raykernels to store rayData which can be later visualized using RayTracingExodus?

**Reply by GiudGiud** [reply]
you can do that yes.

You are visualizing the trajectory of the particles already. It s the same on every time step because the study is run every time step, and the stopping criteria for the rays/particles is not tied to time (the velocity of the particle vs the size of the time step) currently

**Reply by friedmud** [reply]
To add a bit more to this - you can actually change the trajectory of the particle _within_ a time step too.  You can stop the particle as often as you like and change its direction.

**Reply by friedmud** [reply]
To be clear: you will need to write your own `RayKernel` to do any of this trajectory modification.
